# This notebook is about learning data analysis using Python #
## Focus is to prepare, visualize and assess a model to get insigh from the data ##

### Step 1 : importing and doing an initial formatting on the data ###

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# path of data 
path = 'https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DA0101EN/automobileEDA.csv'
df = pd.read_csv(path)
df.head()


symboling  normalized-losses         make aspiration num-of-doors  \
0          3                122  alfa-romero        std          two   
1          3                122  alfa-romero        std          two   
2          1                122  alfa-romero        std          two   
3          2                164         audi        std         four   
4          2                164         audi        std         four   

    body-style drive-wheels engine-location  wheel-base    length  ...  \
0  convertible          rwd           front        88.6  0.811148  ...   
1  convertible          rwd           front        88.6  0.811148  ...   
2    hatchback          rwd           front        94.5  0.822681  ...   
3        sedan          fwd           front        99.8  0.848630  ...   
4        sedan          4wd           front        99.4  0.848630  ...   

   compression-ratio  horsepower  peak-rpm city-mpg highway-mpg    price  \
0                9.0       111.0    5000.0       21          27  13495.0   
1                9.0       111.0    5000.0       21          27  16500.0   
2                9.0       154.0    5000.0       19          26  16500.0   
3               10.0       102.0    5500.0       24          30  13950.0   
4                8.0       115.0    5500.0       18          22  17450.0   

  city-L/100km  horsepower-binned  diesel  gas  
0    11.190476             Medium       0    1  
1    11.190476             Medium       0    1  
2    12.368421             Medium       0    1  
3     9.791667             Medium       0    1  
4    13.055556             Medium       0    1  

[5 rows x 29 columns]

we will now develop several models that will predict the price of the car using the variables or features. This is just an estimate but should give us an objective idea of how much the car should cost.

### Step 2 : Linear regression and multiple linear regression ###

#### a/ Linear regression ####

Simple Linear Regression is a method to help us understand the relationship between two variables:
The predictor/independent variable (X)
The response/dependent variable (that we want to predict)(Y)
The result of Linear Regression is a linear function that predicts the response (dependent) variable as a function of the predictor (independent) variable.

Linear function:
    
𝑌ℎ𝑎𝑡=𝑎+𝑏𝑋

'𝑌' is the r𝑒𝑠𝑝𝑜𝑛𝑠𝑒 𝑉𝑎𝑟𝑖𝑎𝑏𝑙𝑒.  
'𝑋' is the p𝑟𝑒𝑑𝑖𝑐𝑡𝑜𝑟 𝑉𝑎𝑟𝑖𝑎𝑏𝑙𝑒.  
'a' refers to the intercept of the regression line 0, in other words: the value of Y when X is 0.  
'b' refers to the slope of the regression line, in other words: the value with which Y changes when X increases by 1 unit.      

In [2]:
from sklearn.linear_model import LinearRegression

lm = LinearRegression()
lm

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)

In [3]:
# Let's try and example and see how highway-mpg can help us predict car price:

X = df[['highway-mpg']]
Y = df['price']
lm.fit(X,Y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)

In [5]:
# We can output a prediction this way:

Yhat=lm.predict(X)
print (Yhat[0:5])   

# Check the value of the intercept:

print (lm.intercept_)

# or check the value of the slope:

print(lm.coef_)

[16236.50464347 16236.50464347 17058.23802179 13771.3045085
 20345.17153508]
38423.305858157386
[-821.73337832]


So, the model we get is : price = 38423.31 - 821.73 x highway-mpg

In [6]:
# Let's now repeat the same operation with engine size:

X = df[['engine-size']]
Y = df['price']
lm.fit(X,Y)

print (lm.intercept_)
print (lm.coef_)

-7963.338906281049
[166.86001569]


So, the model we get is : price = 166.86 x engine-size -7963.34